Framework/Geometry/test/DetectorInfoIteratorTest.h

```diff
@@ -0,0 +1,181 @@
+the Free Software Foundation; either version 3 of the License, or
+(at your option) any later version.
+
+Mantid is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+File change history is stored at: <https://github.com/mantidproject/mantid>
+Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+
+class DetectorInfoIteratorTest : public CxxTest::TestSuite {
+public:
+  // This pair of boilerplate methods prevent the suite being created statically
+  // This means the constructor isn't called when running other tests
+  static DetectorInfoIteratorTest *createSuite() {
+    return new DetectorInfoIteratorTest();
+  }
+  static void destroySuite(DetectorInfoIteratorTest *suite) { delete suite; }
+
+  DetectorInfoIteratorTest(){};
+
+  std::unique_ptr<Mantid::Geometry::DetectorInfo>
+  create_detector_info_object() {
+
+    // Create a very basic instrument to visit
+    auto visitee = createMinimalInstrument(V3D(0, 0, 0),   // Source position
+                                           V3D(10, 0, 0),  // Sample position
+                                           V3D(11, 0, 0)); // Detector position
+
+    // Add 10 more detectors to the instrument
+    for (int i = 2; i < 12; i++) {
+      Detector *det =
+          new Detector("point-detector", i /*detector id*/, nullptr);
+      det->setPos(V3D(10 + i, 0, 0));
+      std::string num = std::to_string(i);
+      det->setShape(createSphere(0.01 /*1cm*/, V3D(0, 0, 0), num));
+      visitee->add(det);
+      visitee->markAsDetector(det);
+    }
+
+    // Create the instrument visitor
+    InstrumentVisitor visitor(visitee);
+
+    // Return the DetectorInfo object
+    return InstrumentVisitor::makeWrappers(*visitee, nullptr).second;
+  }
+
+  void test_iterator_begin() {
+    // Get the DetectorInfo object
+    auto detectorInfo = create_detector_info_object();
+    auto iter = detectorInfo->begin();
+
+    // Check we start at the correct place
+    TS_ASSERT(iter != detectorInfo->end());
+  }
+
+  void test_iterator_end() {
+    // Get the DetectorInfo object
+    auto detectorInfo = create_detector_info_object();
+    auto iter = detectorInfo->end();
+
+    // Check we start at the correct place
+    TS_ASSERT(iter != detectorInfo->begin());
+  }
+
+  void test_iterator_increment() {
+    // Get the DetectorInfo object
+    auto detectorInfo = create_detector_info_object();
+    auto iter = detectorInfo->begin();
+
+    // Check that we start at the beginning
+    TS_ASSERT(iter == detectorInfo->begin());
+
+    // Increment and check index
+    for (int i = 0; i < 11; ++i) {
+      TS_ASSERT_EQUALS(iter->getIndex(), i);
```

Comment:
I would recomment to not pollute the public interface of a class just for testing purposes. The index is an implementation detail and should actually not be tested. Can you achieve the same by writing a pattern, e.g., to the positions, and comparing that instead?

Reply:
So do you mean loop through the positions and check that they are as expected?

Reply:
Ah okay - I'll change it just now. I'll also do the same for the `SpectrumInfoTest` file.